fl4health/model_bases/autoencoders_base.py

```diff
@@ -0,0 +1,162 @@
+            model_type (AutoEncoderType): The model type should be VARIATIONAL_AE.
+            encoder (nn.Module): Encoder module defined by the user.
+            decoder (nn.Module): Decoder module defined by the user.
+        """
+        super().__init__(model_type, encoder, decoder)
+
+    def encode(self, input: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        mu, logvar = self.encoder(input)
+        return mu, logvar
+
+    def decode(self, latent_vector: torch.Tensor) -> torch.Tensor:
+        output = self.decoder(latent_vector)
+        return output
+
+    def sampling(self, mu: torch.Tensor, logvar: torch.Tensor) -> torch.Tensor:
+        std = torch.exp(0.5 * logvar)
+        eps = torch.randn_like(std)
+        return eps.mul(std).add_(mu)
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        mu, logvar = self.encode(input)
+        z = self.sampling(mu, logvar)
+        output = self.decode(z)
+        # The shape of the flattened_output can be later restored by having the training image shape,
+        # or the decoder structure.
+        flattened_output = output.view(output.shape[0], -1)
+        return torch.cat((logvar, mu, flattened_output), dim=1)
+
+
+class ConditionalVAE(AbstractAE):
+    def __init__(
+        self,
+        model_type: AutoEncoderType,
+        num_conditions: int,
+        encoder: nn.Module,
+        decoder: nn.Module,
+    ) -> None:
+        """Initializes the ConditionalVAE class.
+
+        Args:
+            model_type (AutoEncoderType): The model type should be CONDITIONAL_VAE.
+            num_conditions (int): The total number of conditions, which is used to prepare the condition vector.
+            encoder (nn.Module): Encoder module defined by the user.
+            decoder (nn.Module): Decoder module defined by the user.
+        """
+        super().__init__(model_type, encoder, decoder)
+        self.num_conditions = num_conditions
+
+    def cat_input_condition(self, input: torch.Tensor, condition: torch.Tensor) -> torch.Tensor:
+        input_cond = torch.cat((input, condition), dim=-1)
+        return input_cond
+
+    def one_hot(self, condition: torch.Tensor) -> torch.Tensor:
+        # One-hot encoding of the condition
+        condition = F.one_hot(condition.to(torch.int64), self.num_conditions).to(condition.device)
+        return condition
+
+    def maybe_reshape(self, condition_matrix: torch.Tensor) -> torch.Tensor:
+        # Make sure the condition has the right shape.
+        if len(condition_matrix.shape) > 2:
+            # Since the condition is replicated to match the size of the input,
+            #  we need to narrow it back just to one number per sample
+            return condition_matrix[:, 0, 0]
+        else:
+            # Condition was not replicated to match the size of the input.
+            return condition_matrix
+
+    def encode(self, input: torch.Tensor, condition: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```

Comment:
If I recall correctly, CVAEs admit a few different ways to incorporate conditioning. That is, you can condition both the encoder and the decoder or only one of them. To accommodate this, maybe we make the `condition` tensor Optional and default to None if it isn't passed?